machineconfiguration/v1alpha1/types_machineosconfig.go

```diff
@@ -0,0 +1,224 @@
+	// baseImagePullSecret is the secret used to pull the base image.
+	// must live in the MCO namespace
+	// +kubebuilder:validation:Required
+	BaseImagePullSecret ImageSecretObjectReference `json:"baseImagePullSecret"`
+	// machineOSImageBuilder describes which image builder will be used in each build triggered by this MachineOSConfig
+	// +kubebuilder:validation:Required
+	ImageBuilder *MachineOSImageBuilder `json:"imageBuilder"`
+	// renderedImagePushSecret is the secret used to connect to a user registry.
+	// the final image push and pull secrets should be separate for security concerns. If the final image push secret is somehow exfiltrated,
+	// that gives someone the power to push images to the image repository. By comparison, if the final image pull secret gets exfiltrated,
+	// that only gives someone to pull images from the image repository. It's basically the principle of least permissions.
+	// this push secret will be used only by the MachineConfigController pod to push the image to the final destination. Not all nodes will need to push this image, most of them
+	// will only need to pull the image in order to use it.
+	// +kubebuilder:validation:Required
+	RenderedImagePushSecret ImageSecretObjectReference `json:"renderedImagePushSecret"`
+	// renderedImagePullspec describes the location of the final image.
+	// the MachineOSConfig object will use the in cluster image registry configuration.
+	// if you wish to use a mirror or any other settings specific to registries.conf, please specify those in the cluster wide registries.conf.
+	// The format of the image pushspec is:
+	// host[:port][/namespace]/name:<tag> or svc_name.namespace.svc[:port]/repository/name:<tag>
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MaxLength=447
+	// +kubebuilder:validation:XValidation:rule=`((self.split(':').size() == 2 && self.split(':')[1].matches('^([a-zA-Z0-9-./:])+$')) || self.matches('^[^.]+\\.[^.]+\\.svc:\\d+\\/[^\\/]+\\/[^\\/]+:[^\\/]+$'))`,message="the OCI Image reference must end with a valid :<tag>, where '<digest>' is 64 characters long and '<tag>' is any valid string  Or it must be a valid .svc followed by a port, repository, image name, and tag."
+	// +kubebuilder:validation:XValidation:rule=`((self.split(':').size() == 2 && self.split(':')[0].matches('^([a-zA-Z0-9-]+\\.)+[a-zA-Z0-9-]+(:[0-9]{2,5})?/([a-zA-Z0-9-_]{0,61}/)?[a-zA-Z0-9-_.]*?$')) || self.matches('^[^.]+\\.[^.]+\\.svc:\\d+\\/[^\\/]+\\/[^\\/]+:[^\\/]+$'))`,message="the OCI Image name should follow the host[:port][/namespace]/name format, resembling a valid URL without the scheme. Or it must be a valid .svc followed by a port, repository, image name, and tag."
+	// +kubebuilder:validation:Required
+	RenderedImagePushspec string `json:"renderedImagePushspec"`
+	// releaseVersion is associated with the base OS Image. This is the version of Openshift that the Base Image is associated with.
+	// +optional
```

Comment:
What happens when this isn' specified?  How is this validated?  What format shoudl it be in?  how wil it work pre-release?  What is it used for?

Reply:
added godoc